File: Data Structures & Algorithms/combination-target-sum-ii/submission-0.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        
        candidates.sort()
        path = []
        res = []
        
        
        def dfs(start, curSum):
            if curSum == target:
                res.append(path.copy())
                return
            
            if curSum > target:
                return
            
            for i in range(start, len(candidates)):
                if  i > start and candidates[i] == candidates[i - 1]:
                    continue
                
                path.append(candidates[i])
                dfs(i + 1, curSum + candidates[i])
                path.pop()
            
        
        
        dfs(0, 0)
        return res
            

            

File: Data Structures & Algorithms/combinations/submission-0.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<int> p;
        vector<vector<int>> res;

        backtrack(1, n, k, p, res);

        return res;
    }

    void backtrack(int start, int n, int k, vector<int>& path, vector<vector<int>>& res)
    {
        if(path.size() == k)
        {
            res.push_back(path);
            return;
        }

        for(int i = start; i <= n; i++)
        {
            path.push_back(i);
            backtrack(i + 1, n, k, path, res);
            path.pop_back();
        }
    }
};